### PR TITLE
Optimize TestEventsOOMDisableTrue

### PR DESCRIPTION
**Pull request by WeiZhang555**
Part of #19425 

Improve `TestEventsOOMDisableTrue`: wait less time for OOM.

I have done some research on test time of `TestEventsOOMDisableFalse` to see how long it takes to trigger OOM, basically it cost less than 1 second, except on ARM it cost less than 4.8 seconds, so I think we can make the threshold to 6s.

Fortunately ARM will skip `TestEventsOOMDisableTrue`, so this won't be a problem right now, 6s will be enough for other platforms. If someday ARM start to test on `TestEventsOOMDisableTrue`, 6 second will possibly make it a flaky test, then we have to make it longer or find a better OOM trigger method.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
Perhaps you use the description of this PR as the commit-message? Also, may be worth adding a "TODO" comment to indicate the timeout has to be increased for ARM, once it's enabled (perhaps set a different timeout for ARM than other platforms)

**Comment by WeiZhang555** [comment]
@thaJeztah That's good idea. I've already changed the comment and commit message according to your suggestion. :smile:

**Comment by cpuguy83** [comment]
NOTLGTM

This is a particularly bad use of `<-time.After` and is really racey.
I think it would be best to actually watch the events and wait for an `oom`, and timeout after 20 seconds.

**Comment by WeiZhang555** [comment]
@cpuguy83 Yes, it would.  I didn't change the code logic because it behaved correctly in the past, and according to past test cases, 5 seconds is enough and we don't need 20 seconds.

But I will refactor this test case because it has potential risk as you said.

**Comment by WeiZhang555** [comment]
@cpuguy83 I made some modifications.
Now it will detect OOM events every 2 seconds, if no OOM events monitored, it then waits another 2 seconds until 10 times of retry. So it will wait 20 seconds for OOM to happen totally, and if everything goes very well, this test case may take only 2 seconds to finish.
Please tell me what do you think of this, thank you :smile:

**Comment by cpuguy83** [comment]
Can you have a look at `TestEventsStreaming` and see the usage of the event observer/matcher?
I think this will clean up the code quite a bit.

**Comment by WeiZhang555** [comment]
@cpuguy83 This is a good reference, I have made some changes according to your suggestion, it works really fine, only takes 1 second to finish!

Thank you! Always learn a lot from you! Please take another look, hope I'm doing it right.

File: integration-cli/docker_cli_events_unix_test.go [hunk 1]
```diff
@@ -78,33 +78,42 @@ func (s *DockerSuite) TestEventsOOMDisableFalse(c *check.C) {
...
+	err = observer.Start()
+	c.Assert(err, checker.IsNil)
+	defer observer.Stop()
+
 	go func() {
-		defer close(errChan)
 		out, exitCode, _ := dockerCmdWithError("run", "--oom-kill-disable=true", "--name", "oomTrue", "-m", "10MB", "busybox", "sh", "-c", "x=a; while true; do x=$x$x$x$x; done")
 		if expected := 137; exitCode != expected {
-			errChan <- fmt.Errorf("wrong exit code for OOM container: expected %d, got %d (output: %q)", expected, exitCode, out)
+			c.Fatalf("wrong exit code for OOM container: expected %d, got %d (output: %q)", expected, exitCode, out)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_events_unix_test.go]: This needs to be called from the main goroutine, otherwise the Fatal call won't work.

Comment (WeiZhang555) [inline integration-cli/docker_cli_events_unix_test.go]: Ah, so that's why the check is outside before. I'll change it back.

**Comment by WeiZhang555** [comment]
Updated.

**Comment by cpuguy83** [comment]
LGTM

**Comment by estesp** [comment]
LGTM

**Comment by calavera** [comment]
LGTM

**Pull request merged**